### DISCUSSION TITLE
Is it really incremental parsing?

### DISCUSSION THREAD

**Discussion by dabrahams** [tree-sitter/tree-sitter#3481] [Q&A] [answered]
Some very cool technology here!  I just watched one of the author's presentations; I was interested in the algorithms.  If I understand correctly, the parser is running on the entire input text again, reusing parts of the parse tree from regions that haven't changed when they match the previous parse results.  Is that correct?  If so, I'm a little confused, because he also says that it doesn't take time proportional to the length of the whole input file.  I understand that the constant factor associated with building new nodes is probably very high, but that would not change the fact that every character needs to be lexed.

(If I understood the description correctly, I wouldn't call that incremental parsing… though the complexity of doing actual incremental parsing—which would avoid processing unchanged input text—is high enough that the solution taken by tree-sitter might be a far better one in practice.)

**Comment by maxbrunsfeld** [comment]
Unchanged parts of the input do not need to be lexed.

**Reply by dabrahams** [reply]
Okay, good to know, but then, don't they still need to be parsed?  The alternative would seem to require snapshotting the parser state at each phase of translation so parsing can be resumed.  And also, it seems clear to me that a single token change can cause the remainder of the input to be parsed differently, so in the worst case, even with the most incremental implementation possible, you still need to do work proportional to the length of the input.  What am I missing?

**Reply by dabrahams** [reply]
This is really a separate topic, but for what it's worth, I don't see how you can avoid potentially lexing the whole input in the worst case.  In general even a DFA-based lexer needs to maintain a buffer so it can process input after a long match fails:

```
t0 := .*x
t1 := a
t2 := b
```

The usual algorithm for tokenizing recognizes the longest matching token pattern at each location.  The with the input `"ababababababa"` you have to scan to the end to discover you don't have `t0`.  A change anywhere in the text causes everything to need to be scanned again.

**Reply by maxbrunsfeld** [reply, marked as answer]
Yeah, in the worst case (like commenting / uncommenting the entire file), the entire file needs to be reparsed. But generally, when the user is making edits by typing, you can avoid parsing, lexing, or doing anything with a large portion of the input text, and instead reuse large parts of the previous syntax tree.

**Reply by dabrahams** [reply]
How can it avoid re-parsing? You may need to recover parser states from forked parse stacks that were discarded, it seems to me.

**Reply by dabrahams** [reply]
@maxbrunsfeld Bump?

**Reply by osa1** [reply]
@dabrahams You don't need to save and restore parse states to avoid re-parsing. Example algorithm:

- You create an initial tree by parsing from scratch.
- Every time there's an edit you let the tree know. The tree marks the nodes that can potentially be changed by the edit as "modified" (or "dirty" etc.).
- When you need a new tree after some edits, you start from the beginning but you "shift" unmodified nodes with lookahead that are also not modified, instead of re-parsing. When a node is modified, you start re-parsing but shift sub-nodes that are similarly unmodified with unmodified lookaheads.

There may be other algorithms as well and I'm not sure how tree-sitter works exactly. The point is you don't need to save/restore parse states to be able to reuse nodes and avoid re-parsing unmodified nodes.

As far as I'm aware the most approachable intro to this topic is [this](https://diekmann.uk/diekmann_phd.pdf).

**Reply by dabrahams** [reply]
I'll look at the paper again, but on its face that approach sounds too simplistic. GLR is a nondeterministic parsing algorithm which means it explores multiple possible parses at once. A late failure in one branch of the exploration can cause a different branch, which has little in common with the one that resulted in the tree, to succeed, so the old parse tree after the change may no longer be relevant. That's the reason I think it may be necessary to recover the state of the parser. But maybe the paper has answers...

**Comment by lbguilherme** [comment]
The incremental algorithm runs on top of a list of changes of the input text. Each change has a range and a new text. It doesn't take as input the whole new plain text. This way it can jump straight to the affected nodes. Very clever indeed.
